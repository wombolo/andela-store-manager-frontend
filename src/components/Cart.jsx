import React, { Component } from 'react';
import {connect} from "react-redux";
import PropTypes from "prop-types";
import {getAllCartItems, handleCheckout, removeFromStoreCart} from "../actions/productActions";

export class Cart extends Component {
  state = {};

  componentWillMount() {
    const {cart} = this.props;
    this.setState({cart});
  }

  handleQuantityChange = (event) =>{
    const itemId = parseInt(event.currentTarget.dataset.itemid);
    const itemValue = event.target.value;

    const cart = this.state.cart.splice(0);
    const position =  cart.findIndex((elem) => (elem.id === itemId ));
    cart[position].cartQty = itemValue;
    this.setState({cart})
  };

  handleRemoveFromCart = (id) =>{
    const cart = this.state.cart.splice(0);
    const position =  cart.findIndex((elem) => (elem.id === id ));
    cart.splice(position, 1);

    this.setState({cart: cart});
    this.props.removeFromStoreCart(cart);

  };

  handleSubmit = (event) => {
    event.preventDefault();
    this.props.handleCheckout(this.state.cart);
  };

  render() {
    const {cart} = this.state;
    let iteration = 0, grandTotalItems = 0, grandTotalAmount = 0;

    return (
      <div>
        <h1>Checkout</h1>

        <form onSubmit={this.handleSubmit}>
          <div className="row clearfix">
            <div className="col-10 m-0-auto m-t-5">
              <h3>Cart </h3>

              <table className="dataTable table-striped" border="1" cellPadding="5">
                <thead>
                <tr>
                  <td>S/N</td>
                  <td>Transaction ID</td>
                  <td>Item Image</td>
                  <td>Item Name</td>
                  <td>Item Price</td>
                  <td>Item Quantity</td>
                  <td>Amount</td>
                  <td>Action</td>
                </tr>
                </thead>

                <tbody>
                {cart.length > 0 ?
                  cart.map(item =>{
                  const thisAmountTotal = parseFloat((item.price * item.cartQty).toFixed(2));

                  iteration++;
                  grandTotalItems += parseInt(item.cartQty);
                  grandTotalAmount += thisAmountTotal;

                  return(
                    <tr key={iteration}>
                    <td>{iteration}</td>
                    <td><code>{Math.random().toString(36).substring(7).toUpperCase()}</code></td>
                    <td><img src={require(`../assets/images/${item.image}`)} /></td>
                    <td>{item.title}</td>
                    <td>$ {item.price}</td>
                    <td><input type={'number'}
                               data-itemid={item.id}
                               defaultValue={item.cartQty}
                               max={item.cartQty + item.quantity}
                               onChange={this.handleQuantityChange} /></td>
                    <td>$ {thisAmountTotal}</td>
                    <td><button type='reset' id={`remove-${item.id}`} className='text-danger' onClick={()=> (this.handleRemoveFromCart(item.id))}>Remove</button></td>
                  </tr>
                  )
                })
                  : <tr><td colSpan="8" className='center'>No Items in Cart</td></tr>}
                </tbody>

                <tfoot>
                <tr>
                  <td colSpan="5">Grand Total</td>
                  <td>{grandTotalItems}</td>
                  <td>$ {grandTotalAmount}</td>
                  <td> </td>
                </tr>
                </tfoot>
              </table>
            </div>
          </div>

          {cart.length > 0 ?
            <div className="row clearfix">
              <div className="col-8"></div>
              <div className="col-2 pull-right">
                <button className="btn btn-success">Checkout</button>
              </div>
            </div>
            : ''
          }
        </form>
      </div>
    )
  }
}

Cart.propTypes = {
  getAllCartItems: PropTypes.func.isRequired,
  handleCheckout: PropTypes.func.isRequired,
  removeFromStoreCart: PropTypes.func.isRequired,
  cart: PropTypes.array.isRequired,
  profile: PropTypes.shape({}).isRequired,
};

export const mapStateToProps = state => ({
  cart: state.products.cart,
  profile: state.auth.user.profile
});

export default connect(mapStateToProps,{
  getAllCartItems, handleCheckout, removeFromStoreCart
})
(Cart);
