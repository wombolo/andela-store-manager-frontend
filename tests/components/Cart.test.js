import 'babel-polyfill'
import React from 'react';
import { mount, shallow } from 'enzyme';
import sinon from 'sinon';
import { MemoryRouter } from 'react-router-dom';

import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';

import { Cart } from '../../src/components/Cart.jsx';

let wrapper;

const props = {
  getAllCartItems: () => Promise.resolve(),
  handleCheckout: () => Promise.resolve(),
  cart:[
    {id: 1, title: "Bonjour", image: "products/product-5.png",price: 150, quantity: 3, cartQty: '3', status: "active"},
    {id: 2, title: "Bonjour", image: "products/product-5.png",price: 150, quantity: 3, cartQty: '3', status: "active"},
  ],
  profile: {
    id: 1,
    firstname: 'Akano',
    lastname: 'Peter',
    email: 'elkenzom1@gmail.com',
    role: 'admin',
    image: 'team-7.jpg',
  },
};


const mockStore = configureStore([thunk]);

const store = mockStore({ ...props });



describe('Cart Component', () => {
  it('should match snapshot', () => {
    wrapper = shallow(<Cart {...props}/>);
    expect(wrapper).toMatchSnapshot();
  });

  describe('Methods in Class', () => {
    beforeEach(() => {
      wrapper = mount(
        <Provider store={store}>
          <MemoryRouter>
          <Cart {...props} />
          </MemoryRouter>
        </Provider>,
      );
    });

    afterEach(() => {
      wrapper.unmount();
    });

    it('triggers handleQuantityChange()', () => {
      const CartComp = wrapper.find('Cart')
      sinon.spy(CartComp.instance(), 'handleQuantityChange');

      CartComp.instance().handleQuantityChange({
        currentTarget:{dataset:{itemid: 2}},
        target:{value:'Hiii Book'}
      });
      expect(CartComp.instance().handleQuantityChange.calledOnce)
        .toEqual(true);
    });


    it('triggers handleRemoveFromCart()', () => {
      const CartComp = wrapper.find('Cart');

      sinon.spy(CartComp.instance(), 'handleRemoveFromCart');

      CartComp.instance().handleRemoveFromCart(1);
      expect(CartComp.instance().handleRemoveFromCart.calledOnce)
        .toEqual(true);

      // const button = CartComp.find('button').last();
      // button.simulate('click', 1);
    });


    it('triggers handleSubmit()', () => {
      const CartComp = wrapper.find('Cart');

      sinon.spy(CartComp.instance(), 'handleSubmit');
      CartComp.instance().handleSubmit({preventDefault: () => {}});
      expect(CartComp.instance().handleSubmit.calledOnce)
        .toEqual(true);
    });
  });
});
