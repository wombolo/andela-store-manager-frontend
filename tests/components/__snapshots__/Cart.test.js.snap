// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart Component should match snapshot 1`] = `
<div>
  <h1>
    Checkout
  </h1>
  <form
    onSubmit={[Function]}
  >
    <div
      className="row clearfix"
    >
      <div
        className="col-10 m-0-auto m-t-5"
      >
        <h3>
          Cart 
        </h3>
        <table
          border="1"
          cellPadding="5"
          className="dataTable table-striped"
        >
          <thead>
            <tr>
              <td>
                S/N
              </td>
              <td>
                Transaction ID
              </td>
              <td>
                Item Image
              </td>
              <td>
                Item Name
              </td>
              <td>
                Item Price
              </td>
              <td>
                Item Quantity
              </td>
              <td>
                Amount
              </td>
              <td>
                Action
              </td>
            </tr>
          </thead>
          <tbody>
            <tr
              key="1"
            >
              <td>
                1
              </td>
              <td>
                <code>
                  U804B
                </code>
              </td>
              <td>
                <img
                  src=""
                />
              </td>
              <td>
                Bonjour
              </td>
              <td>
                $ 
                150
              </td>
              <td>
                <input
                  data-itemid={1}
                  defaultValue="3"
                  max="33"
                  onChange={[Function]}
                  type="number"
                />
              </td>
              <td>
                $ 
                450
              </td>
              <td>
                <button
                  className="text-danger"
                  id="remove-1"
                  onClick={[Function]}
                  type="reset"
                >
                  Remove
                </button>
              </td>
            </tr>
            <tr
              key="2"
            >
              <td>
                2
              </td>
              <td>
                <code>
                  IWJCF2
                </code>
              </td>
              <td>
                <img
                  src=""
                />
              </td>
              <td>
                Bonjour
              </td>
              <td>
                $ 
                150
              </td>
              <td>
                <input
                  data-itemid={2}
                  defaultValue="3"
                  max="33"
                  onChange={[Function]}
                  type="number"
                />
              </td>
              <td>
                $ 
                450
              </td>
              <td>
                <button
                  className="text-danger"
                  id="remove-2"
                  onClick={[Function]}
                  type="reset"
                >
                  Remove
                </button>
              </td>
            </tr>
          </tbody>
          <tfoot>
            <tr>
              <td
                colSpan="5"
              >
                Grand Total
              </td>
              <td>
                6
              </td>
              <td>
                $ 
                900
              </td>
              <td>
                 
              </td>
            </tr>
          </tfoot>
        </table>
      </div>
    </div>
    <div
      className="row clearfix"
    >
      <div
        className="col-8"
      />
      <div
        className="col-2 pull-right"
      >
        <button
          className="btn btn-success"
        >
          Checkout
        </button>
      </div>
    </div>
  </form>
</div>
`;
